services/sharedchannel/service.go

```diff
@@ -0,0 +1,155 @@
+	UpdatePost(post *model.Post, safeUpdate bool) (*model.Post, *model.AppError)
+}
+
+// Service provides shared channel synchronization.
+type Service struct {
+	server       ServerIface
+	app          AppIface
+	changeSignal chan struct{}
+
+	// everything below guarded by `mux`
+	mux                   sync.RWMutex
+	active                bool
+	leaderListenerId      string
+	done                  chan struct{}
+	tasks                 map[string]syncTask
+	syncTopicListenerId   string
+	inviteTopicListenerId string
+}
+
+// NewSharedChannelService creates a RemoteClusterService instance.
+func NewSharedChannelService(server ServerIface, app AppIface) (*Service, error) {
+	service := &Service{
+		server:       server,
+		app:          app,
+		changeSignal: make(chan struct{}, 1),
+		tasks:        make(map[string]syncTask),
+	}
+	return service, nil
+}
+
+// Start is called by the server on server start-up.
+func (scs *Service) Start() error {
+	scs.mux.Lock()
+	scs.leaderListenerId = scs.server.AddClusterLeaderChangedListener(scs.onClusterLeaderChange)
+	scs.mux.Unlock()
+
+	scs.onClusterLeaderChange()
+
+	return nil
+}
+
+// Shutdown is called by the server on server shutdown.
+func (scs *Service) Shutdown() error {
+	scs.server.RemoveClusterLeaderChangedListener(scs.leaderListenerId)
+	scs.pause()
+	return nil
+}
+
+func (scs *Service) sendEphemeralPost(channelId string, userId string, text string) {
+	ephemeral := &model.Post{
+		ChannelId: channelId,
+		Message:   text,
+		CreateAt:  model.GetMillis(),
+	}
+	scs.app.SendEphemeralPost(userId, ephemeral)
+}
+
+// onClusterLeaderChange is called whenever the cluster leader may have changed.
+func (scs *Service) onClusterLeaderChange() {
+	if scs.server.IsLeader() {
+		scs.resume()
+	} else {
+		scs.pause()
+	}
+}
+
+func (scs *Service) resume() {
+	scs.mux.Lock()
+	defer scs.mux.Unlock()
+
+	if scs.active {
+		return // already active
+	}
+
+	rcs := scs.server.GetRemoteClusterService()
+	if rcs == nil {
+		scs.server.GetLogger().Error("Shared Channel Service cannot activate: requires Remote Cluster Service")
+		return
+	}
+	scs.syncTopicListenerId = rcs.AddTopicListener(TopicSync, scs.onReceiveSyncMessage)
+	scs.inviteTopicListenerId = rcs.AddTopicListener(TopicChannelInvite, scs.onReceiveChannelInvite)
+
+	scs.active = true
+	scs.done = make(chan struct{})
+
+	scs.syncLoop(scs.done)
+
+	scs.server.GetLogger().Debug("Shared Channel Service active")
+}
+
+func (scs *Service) pause() {
+	scs.mux.Lock()
+	defer scs.mux.Unlock()
+
+	if !scs.active {
+		return // already inactive
+	}
+
+	rcs := scs.server.GetRemoteClusterService()
+	if rcs == nil {
+		scs.server.GetLogger().Error("Shared Channel Service activitate: requires Remote Cluster Service")
+	}
+	rcs.RemoveTopicListener(scs.syncTopicListenerId)
+	scs.syncTopicListenerId = ""
+	rcs.RemoveTopicListener(scs.inviteTopicListenerId)
+	scs.inviteTopicListenerId = ""
+
+	scs.active = false
+	close(scs.done)
```

Comment:
Are we okay with not waiting for `syncLoop` to be actually done?

Reply:
Yes, if it is in the middle of a sync batch then I want it to gracefully finish what it was working on then exit.  There are already timeouts built-in to ensure this is a bounded time-to-live.